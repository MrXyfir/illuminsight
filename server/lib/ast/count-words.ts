import { Illuminsight } from 'types/illuminsight';

/**
 * Count words in AST node.
 */
export function countWords(node: Illuminsight.AST): number {
  // Count words in string
  if (typeof node == 'string') {
    const text = node.trim();
    return !text ? 0 : text.split(/\s+|\-+/).length;
  }
  // Count words recursively in child nodes
  else if (node.c) {
    let words = 0;
    for (let child of node.c) words += countWords(child);
    return words;
  }
  return 0;
}
