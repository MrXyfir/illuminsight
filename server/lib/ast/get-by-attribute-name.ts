import { Illuminsight } from 'types/illuminsight';

/**
 * Get nodes from AST by attribute name.
 */
export function getByAttributeName(
  name: string,
  ast: Illuminsight.AST[]
): Illuminsight.AST[] {
  const matches: Illuminsight.AST[] = [];

  for (let node of ast) {
    // Ignore string nodes
    if (typeof node == 'string') continue;
    // Check attribute
    else {
      // Include node if it matches
      if (node.a && node.a[name] !== undefined) matches.push(node);

      // Include chilren if they match
      if (node.c) matches.push(...getByAttributeName(name, node.c));
    }
  }

  return matches;
}
