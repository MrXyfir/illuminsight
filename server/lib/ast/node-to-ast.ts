import { Insightful } from 'types/insightful';

const REPLACEMENT_ATTRIBUTES: { [attr: string]: string } = {
  'xlink:href': 'href'
};
const EXCLUDED_ATTRIBUTES = ['class', 'tag'];
const EXCLUDED_ELEMENTS = ['iframe', 'script', 'style', 'link'];

/**
 * Convert a DOM node to AST.
 */
export function nodeToAST(
  node: Node,
  /**
   * Preserve whitespace if a child of a `<pre>` element.
   */
  pre: boolean = false
): Insightful.AST | undefined {
  // Element node
  if (node.nodeType == node.ELEMENT_NODE) {
    const ast: Insightful.AST = { n: node.nodeName.toLowerCase() };

    // Ignore excluded elements
    if (EXCLUDED_ELEMENTS.includes(ast.n)) return;

    // Copy all but excluded attributes
    for (let attr of (node as Element).attributes) {
      if (EXCLUDED_ATTRIBUTES.includes(attr.name)) continue;

      const name = REPLACEMENT_ATTRIBUTES[attr.name] || attr.name;
      if (ast.a) ast.a[name] = attr.value;
      else ast.a = { [name]: attr.value };
    }

    // Recursively build AST for child nodes
    for (let childNode of node.childNodes) {
      const childAST = nodeToAST(childNode, pre || ast.n == 'pre');
      if (childAST) {
        // Add child to list
        if (ast.c) ast.c.push(childAST);
        // Create list with child
        else ast.c = [childAST];
      }
    }

    return ast;
  }
  // Text or CDATA node (which we treat same as text)
  else if (
    (node.nodeType == node.TEXT_NODE ||
      node.nodeType == node.CDATA_SECTION_NODE) &&
    node.textContent
  ) {
    // Preserve text content as is without manipulating whitespace
    if (pre) return node.textContent;

    // Convert newlines and multiple spaces to a single space
    const text = node.textContent.replace(/\n/g, ' ').replace(/\s{2,}/g, ' ');

    // Only return if it's more than just a single space
    if (text != ' ') return text;
  }
}
