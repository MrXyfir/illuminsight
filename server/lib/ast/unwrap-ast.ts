import { Insightful } from 'types/insightful';

/**
 * Remove top-level wrapper elements that will prevent bookmark's "blocks"
 *  system from working.
 */
export function unwrapAST(node: Insightful.AST): Insightful.AST[] {
  // We received a string (shouldn't happen)
  if (typeof node == 'string') return [node];

  // Node does not have children (shouldn't happen)
  if (!node.c || !node.c.length) return [node];

  // Unwrap this node's only child
  if (node.c.length == 1 && typeof node.c[0] != 'string')
    return unwrapAST(node.c[0]);

  // Return node's non-string, non-wrapper children
  if (!node.c.some(c => typeof c == 'string')) return node.c;

  // Node contains string children, so keep it
  return [node];
}
