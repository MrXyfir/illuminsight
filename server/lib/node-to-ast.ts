import { Insightful } from 'types/insightful';

/**
 * Convert a DOM node to AST.
 */
export function nodeToAST(node: Node): Insightful.AST | undefined {
  // Element node
  if (node.nodeType == node.ELEMENT_NODE) {
    const ast: Insightful.AST = { n: node.nodeName.toLowerCase() };

    // Save link's href attribute
    if (ast.n == 'a') {
      ast.a = { href: (node as HTMLAnchorElement).href };
    }
    // Save image's src attribute
    if (ast.n == 'img') {
      const { src, alt } = node as HTMLImageElement;
      ast.a = { src };
      if (alt) ast.a.alt = alt;
    }

    // Recursively build AST for child nodes
    for (let childNode of node.childNodes) {
      const childAST = nodeToAST(childNode);
      if (childAST) {
        // Add child to list
        if (ast.c) ast.c.push(childAST);
        // Create list with child
        else ast.c = [childAST];
      }
    }

    return ast;
  }
  // Text or CDATA node (which we treat same as text)
  else if (
    (node.nodeType == node.TEXT_NODE ||
      node.nodeType == node.CDATA_SECTION_NODE) &&
    node.textContent
  ) {
    // Only return if it still has content after trimming
    const text = node.textContent.trim();
    if (text) return text;
  }
}
