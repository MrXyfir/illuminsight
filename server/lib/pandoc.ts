import { exec } from 'child_process';

export function pandoc({
  output,
  input,
  from,
  to,
  ...opt
}: {
  output: string;
  input: string;
  from: string;
  to: string;
  [x: string]: string;
}): Promise<void> {
  const options = Object.entries(opt)
    .map(o => `--${o[0]}="${o[1]}"`)
    .join(' ');
  return new Promise((resolve, reject) => {
    exec(
      `pandoc -f ${from} -t ${to} -o "${output}" "${input}" ${options}`,
      (err, stdout, stderr) => {
        if (err || stderr) reject(err || stderr);
        resolve();
      }
    );
  });
}
