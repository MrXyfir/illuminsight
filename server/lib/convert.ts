import { Insightful } from 'types/insightful';
import { countWords } from 'lib/count-words';
import { nodeToAst } from 'lib/node-to-ast';
import * as archiver from 'archiver';
import { Calibre } from 'node-calibre';
import { resolve } from 'path';
import { Extract } from 'unzipper';
import { pandoc } from 'lib/pandoc';
import { JSDOM } from 'jsdom';
import {
  createWriteStream,
  createReadStream,
  ReadStream,
  writeJSON,
  writeFile,
  readFile,
  remove,
  mkdir,
  move
} from 'fs-extra';

const calibre = new Calibre({ log: process.enve.NODE_ENV == 'development' });

export async function convert({
  file,
  link,
  text
}: {
  file?: string;
  link?: string;
  text?: string;
}): Promise<ReadStream> {
  // Save to .txt file
  if (text) {
    file = resolve(process.enve.TEMP_DIR, `${Date.now()}.txt`);
    await writeFile(file, text);
    text = undefined;
  }

  // Create EPUB from webpage
  if (link) {
    // Set file and directory paths we'll use later
    const webpageDirectory = resolve(
      process.enve.TEMP_DIR,
      `webpage-${Date.now()}`
    );
    const webpageImgDirectory = resolve(webpageDirectory, 'images');
    const markdownFile = resolve(webpageDirectory, 'index.md');
    const htmlFile = resolve(webpageDirectory, 'index.html');

    // Pandoc converts page to CommonMark-raw_html to discard unwanted elements
    await pandoc({
      'extract-media': webpageImgDirectory,
      output: markdownFile,
      input: link,
      from: 'html',
      to: 'commonmark-raw_html'
    });

    // Convert Markdown back to HTML
    await pandoc({
      output: htmlFile,
      input: markdownFile,
      from: 'commonmark',
      to: 'html'
    });

    // Convert HTML to EPUB
    file = resolve(process.enve.TEMP_DIR, `${Date.now()}.epub`);
    await pandoc({ output: file, input: htmlFile, from: 'html', to: 'epub' });

    // Delete webpage directory
    await remove(webpageDirectory);
  }

  if (!file) throw new Error('Bad or missing input');

  // Convert to EPUB
  // Even if already an EPUB, it will validate and rebuild as we expect it
  const originalFile = file;
  file = await calibre.ebookConvert(file, 'epub');

  // Extract files from EPUB
  const epubDirectory = resolve(process.enve.TEMP_DIR, `unzip-${Date.now()}`);
  await new Promise(resolve =>
    createReadStream(file as string).pipe(
      Extract({ path: epubDirectory }).on('close', resolve)
    )
  );

  // Delete converted EPUB file and original source
  await remove(originalFile);
  await remove(file);

  // Parse OPF with jsdom
  const opfDom = new JSDOM(
    await readFile(resolve(epubDirectory, 'content.opf')),
    { contentType: 'text/xml' }
  );
  const opfDoc = opfDom.window.document;

  // Create directories for unzipped astpub
  const astpubDirectory = resolve(
    process.enve.TEMP_DIR,
    `astpub-${Date.now()}`
  );
  const astpubImgDirectory = resolve(astpubDirectory, 'images');
  await mkdir(astpubDirectory);
  await mkdir(astpubImgDirectory);

  // Hash map for item id->href
  const idToHref: { [id: string]: string } = {};

  // Relative file path/name for cover image
  const covers: { id1?: string; id2?: string; href?: string } = {};
  let cover: Insightful.Entity['cover'];

  // How many words are in the content
  let words = 0;

  // Loop through package>manifest>item elements
  for (let item of opfDoc.getElementsByTagName('item')) {
    const mediaType = item.getAttribute('media-type');
    const href = item.getAttribute('href');
    const id = item.getAttribute('id');
    if (!mediaType || !href || !id) continue;

    idToHref[id] = href;

    // Move images from EPUB directory to astpub directory
    if (mediaType.startsWith('image/')) {
      await move(
        resolve(epubDirectory, href),
        resolve(astpubImgDirectory, href)
      );

      // Search for href of cover image
      if (!cover) {
        // Guaranteed to be cover image
        const properties = item.getAttribute('properties');
        if (properties == 'cover-image') cover = href;
        // Most likely the cover but a guaranteed option may still be available
        else if (id == 'cover') covers.id1 = href;
        else if (id == 'ci') covers.id2 = href;
        else if (href.startsWith('cover')) covers.href = href;
      }
    }
    // Convert XHTML to our JSON via jsdom
    if (/xhtml|html|xml/.test(mediaType)) {
      // Read file
      const xhtmlDom = new JSDOM(await readFile(resolve(epubDirectory, href)), {
        contentType: 'text/xml'
      });
      const xhtmlDoc = xhtmlDom.window.document;

      // Check for body element since document could be XML (like toc.ncx)
      if (!xhtmlDoc.body) continue;

      // Convert document starting at body to AST
      const ast = nodeToAst(xhtmlDoc.body) as Insightful.AST;

      // Count words in tree
      words += countWords(ast);

      // Write AST to file
      await writeJSON(resolve(astpubDirectory, `${href}.json`), ast);
    }
  }

  // Use fallbacks for cover image if available
  if (!cover) cover = covers.id1 || covers.id2 || covers.href;

  // Delete extracted epub directory
  await remove(epubDirectory);

  // Populate entity object which will be used for meta.json
  const entity: Insightful.Entity = {
    authors:
      Array.from(opfDoc.getElementsByTagName('dc:creator'))
        .map(creator => creator.textContent)
        .join(', ') || undefined,
    bookmark: { element: 0, section: 0, width: 0, line: 0 },
    cover: `images/${cover}`,
    id: Date.now(),
    link,
    name: opfDoc.getElementsByTagName('dc:title')[0].textContent || '',
    published: (() => {
      const date = opfDoc.getElementsByTagName('dc:date')[0];
      if (date) return new Date(date.textContent as string).getTime();
    })(),
    publisher: (() => {
      const pub = opfDoc.getElementsByTagName('dc:publisher')[0];
      if (pub) return pub.textContent as string;
    })(),
    // Order content items for spine
    spine: Array.from(opfDoc.getElementsByTagName('itemref')).map(
      ref => idToHref[ref.getAttribute('idref') as string]
    ),
    starred: false,
    tags: [],
    version: process.enve.ASTPUB_VERSION,
    words:
      words > 999999
        ? `${(words / 1000000).toFixed(2)}m`
        : words > 999
        ? `${Math.round(words / 1000)}k`
        : words.toString()
  };

  // Write meta.json
  await writeJSON(resolve(astpubDirectory, 'meta.json'), entity);

  // Zip directory
  const astpubFile = resolve(process.enve.TEMP_DIR, `astpub-${Date.now()}.zip`);
  const astpubWriter = createWriteStream(astpubFile);
  const astpubArchive = archiver('zip');
  const astpubPromise = new Promise(r => astpubWriter.on('close', r));
  astpubArchive.pipe(astpubWriter);
  astpubArchive.directory(astpubDirectory, false);
  astpubArchive.finalize();
  await astpubPromise;

  // Delete unzipped astpub directory
  await remove(astpubDirectory);

  // Create stream to return for astpub file
  const astpubReader = createReadStream(astpubFile);

  // Delete file after it's been consumed
  astpubReader.on('close', () => remove(astpubFile, () => undefined));

  // Return string to astpub file
  return astpubReader;
}
