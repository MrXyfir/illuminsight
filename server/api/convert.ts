import { NextFunction, Response, Request } from 'express';
import { convert } from 'lib/convert';

export function api_convert(
  req: Request,
  res: Response,
  next: NextFunction
): void {
  const opt = req.file ? { file: req.file.path } : req.body;
  delete req.file;
  convert(opt)
    .then(file => file.pipe(res.status(200)))
    .catch(next);
}
