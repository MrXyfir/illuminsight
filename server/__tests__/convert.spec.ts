import { Illuminsight } from 'types/illuminsight';
import { nodeToAST } from 'lib/ast/node-to-ast';
import { convert } from 'lib/convert';
import { Extract } from 'unzipper';
import { resolve } from 'path';
import { pandoc } from 'lib/pandoc';
import { JSDOM } from 'jsdom';
import {
  constants as FS,
  ensureDir,
  emptyDir,
  readJSON,
  readFile,
  access,
  copy
} from 'fs-extra';
import 'jest-extended';

// Extract environment variables
const { FILES_DIRECTORY, TEMP_DIR } = process.enve;

// Set paths used in tests
const convertDirectory = resolve(TEMP_DIR, 'convert-test');
const astpubDirectory = resolve(convertDirectory, 'astpub');
const astpubMetaFile = resolve(astpubDirectory, 'meta.json');
const loremHtmlFile = resolve(FILES_DIRECTORY, 'lorem.html');
const loremMdFile = resolve(convertDirectory, 'lorem.md');

// Ensure test directory exists and is empty
beforeEach(async () => {
  await ensureDir(convertDirectory);
  await emptyDir(convertDirectory);
});

test('nodeToAST()', async () => {
  const dom = new JSDOM(await readFile(loremHtmlFile));
  expect(nodeToAST(dom.window.document.body)).toMatchSnapshot();
});

test('pandoc()', async () => {
  // Convert HTML to CommonMark
  await pandoc({
    output: loremMdFile,
    input: loremHtmlFile,
    from: 'html',
    to: 'commonmark-raw_html'
  });

  // Read Markdown content
  const markdown = await readFile(loremMdFile, 'utf8');

  // Validate content was converted properly
  expect(markdown).toMatch(/^# Lorem Ipsum/m);
  expect(markdown).toMatch(/^#### "Neque porro/m);
  expect(markdown).toMatch(/^Lorem ipsum dolor/m);
});

test('convert({text})', async () => {
  // Convert content to astpub format then extract
  const readStream = await convert({ text: ' Hello \n  World! How are you? ' });
  await new Promise(resolve =>
    readStream.pipe(Extract({ path: astpubDirectory }).on('close', resolve))
  );

  // Validate meta.json
  const pub: Illuminsight.Pub = await readJSON(astpubMetaFile);
  const _pub: Illuminsight.Pub = {
    authors: 'Unknown',
    bookmark: { section: 0, element: 0 },
    cover: 'res/0.jpg',
    id: pub.id,
    name: 'Hello...',
    toc: [{ section: 0, element: 0, title: 'Start' }],
    sections: 2,
    starred: false,
    tags: [],
    version: 1,
    words: '5'
  };
  expect(pub).toMatchObject(_pub);

  // Validate cover
  await access(resolve(astpubDirectory, 'res/0.jpg'), FS.F_OK);

  // Validate AST
  const ast: Illuminsight.AST[] = await readJSON(
    resolve(astpubDirectory, 'ast/1.json')
  );
  const _ast: Illuminsight.AST[] = [
    { n: 'p', c: ['Hello'] },
    { n: 'p', c: ['World! How are you?'] }
  ];
  expect(ast).toMatchObject(_ast);
});

test(
  'convert({link})',
  async () => {
    // Convert content to astpub format then extract
    const readStream = await convert({
      link:
        'https://www.nytimes.com/2019/05/01/magazine/ehren-tool-war-cups-smithsonian.html'
    });
    await new Promise(resolve =>
      readStream.pipe(Extract({ path: astpubDirectory }).on('close', resolve))
    );

    // Validate meta.json
    const pub: Illuminsight.Pub = await readJSON(astpubMetaFile);
    const _pub: Illuminsight.Pub = {
      ...pub,
      authors: 'Unknown',
      bookmark: { section: 0, element: 0 },
      cover: 'res/0.jpg',
      link:
        'https://www.nytimes.com/2019/05/01/magazine/ehren-tool-war-cups-smithsonian.html',
      name:
        'The Price of This Artist’s Work? A Conversation About the Horrors of War - The New York Times',
      toc: [
        {
          section: 2,
          element: 0,
          title:
            'The Price of This Artist’s Work? A Conversation About the Horrors of War - The New York Times'
        }
      ],
      sections: 4,
      starred: false,
      tags: [],
      version: 1,
      words: '3k'
    };
    expect(pub).toMatchObject(_pub);

    // Validate cover
    await access(resolve(astpubDirectory, 'res/0.jpg'), FS.F_OK);

    // Validate AST
    const ast: Illuminsight.AST[] = await readJSON(
      resolve(astpubDirectory, 'ast/3.json')
    );
    expect(ast).toMatchSnapshot();
  },
  30 * 1000
);

test(
  'convert({file})',
  async () => {
    // Move test ebook to temp_dir
    const file = resolve(TEMP_DIR, `${Date.now()}.epub`);
    await copy(resolve(FILES_DIRECTORY, 'ebook.epub'), file);

    // Convert content to astpub format then extract
    const readStream = await convert({ file });
    await new Promise(resolve =>
      readStream.pipe(Extract({ path: astpubDirectory }).on('close', resolve))
    );

    // Validate meta.json
    const pub: Illuminsight.Pub = await readJSON(astpubMetaFile);
    const _pub: Illuminsight.Pub = {
      ...pub,
      authors: 'Charles Dickens',
      bookmark: { section: 0, element: 0 },
      cover: 'res/0.jpg',
      name: 'A Tale of Two Cities',
      published: 757411200000,
      sections: 54,
      starred: false,
      tags: [],
      version: 1,
      words: '140k'
    };
    expect(pub).toMatchObject(_pub);

    expect(pub.toc).toBeArrayOfSize(50);
    expect(pub.toc.slice(0, 2)).toMatchObject([
      { element: 'pgepubid00000', section: 1, title: 'A TALE OF TWO CITIES' },
      { element: 0, section: 2, title: 'A STORY OF THE FRENCH REVOLUTION' }
    ] as Illuminsight.Pub['toc']);

    // Validate cover
    await access(resolve(astpubDirectory, 'res/0.jpg'), FS.F_OK);

    // Validate AST
    const ast: Illuminsight.AST[] = await readJSON(
      resolve(astpubDirectory, 'ast/5.json')
    );
    expect(ast).toMatchSnapshot();
  },
  30 * 1000
);
