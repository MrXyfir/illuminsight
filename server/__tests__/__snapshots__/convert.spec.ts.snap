// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert({file}) 1`] = `
Array [
  Object {
    "a": Object {
      "id": "pgepubid00003",
    },
    "c": Array [
      "I. The Period",
    ],
    "n": "h2",
  },
  Object {
    "c": Array [
      Object {
        "c": Array [
          Object {
            "c": Array [
              "It was the best of times,",
              Object {
                "n": "br",
              },
              " it was the worst of times,",
              Object {
                "n": "br",
              },
              " it was the age of wisdom,",
              Object {
                "n": "br",
              },
              " it was the age of foolishness,",
              Object {
                "n": "br",
              },
              " it was the epoch of belief,",
              Object {
                "n": "br",
              },
              " it was the epoch of incredulity,",
              Object {
                "n": "br",
              },
              " it was the season of Light,",
              Object {
                "n": "br",
              },
              " it was the season of Darkness,",
              Object {
                "n": "br",
              },
              " it was the spring of hope,",
              Object {
                "n": "br",
              },
              " it was the winter of despair,",
            ],
            "n": "p",
          },
        ],
        "n": "div",
      },
    ],
    "n": "blockquote",
  },
  Object {
    "c": Array [
      "we had everything before us, we had nothing before us, we were all going direct to Heaven, we were all going direct the other way— in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      "There were a king with a large jaw and a queen with a plain face, on the throne of England; there were a king with a large jaw and a queen with a fair face, on the throne of France. In both countries it was clearer than crystal to the lords of the State preserves of loaves and fishes, that things in general were settled for ever.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      "It was the year of Our Lord one thousand seven hundred and seventy-five. Spiritual revelations were conceded to England at that favoured period, as at this. Mrs. Southcott had recently attained her five-and-twentieth blessed birthday, of whom a prophetic private in the Life Guards had heralded the sublime appearance by announcing that arrangements were made for the swallowing up of London and Westminster. Even the Cock-lane ghost had been laid only a round dozen of years, after rapping out its messages, as the spirits of this very year last past (supernaturally deficient in originality) rapped out theirs. Mere messages in the earthly order of events had lately come to the English Crown and People, from a congress of British subjects in America: which, strange to relate, have proved more important to the human race than any communications yet received through any of the chickens of the Cock-lane brood.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      "France, less favoured on the whole as to matters spiritual than her sister of the shield and trident, rolled with exceeding smoothness down hill, making paper money and spending it. Under the guidance of her Christian pastors, she entertained herself, besides, with such humane achievements as sentencing a youth to have his hands cut off, his tongue torn out with pincers, and his body burned alive, because he had not kneeled down in the rain to do honour to a dirty procession of monks which passed within his view, at a distance of some fifty or sixty yards. It is likely enough that, rooted in the woods of France and Norway, there were growing trees, when that sufferer was put to death, already marked by the Woodman, Fate, to come down and be sawn into boards, to make a certain movable framework with a sack and a knife in it, terrible in history. It is likely enough that in the rough outhouses of some tillers of the heavy lands adjacent to Paris, there were sheltered from the weather that very day, rude carts, bespattered with rustic mire, snuffed about by pigs, and roosted in by poultry, which the Farmer, Death, had already set apart to be his tumbrils of the Revolution. But that Woodman and that Farmer, though they work unceasingly, work silently, and no one heard them as they went about with muffled tread: the rather, forasmuch as to entertain any suspicion that they were awake, was to be atheistical and traitorous.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      "In England, there was scarcely an amount of order and protection to justify much national boasting. Daring burglaries by armed men, and highway robberies, took place in the capital itself every night; families were publicly cautioned not to go out of town without removing their furniture to upholsterers' warehouses for security; the highwayman in the dark was a City tradesman in the light, and, being recognised and challenged by his fellow-tradesman whom he stopped in his character of “the Captain,” gallantly shot him through the head and rode away; the mail was waylaid by seven robbers, and the guard shot three dead, and then got shot dead himself by the other four, “in consequence of the failure of his ammunition:” after which the mail was robbed in peace; that magnificent potentate, the Lord Mayor of London, was made to stand and deliver on Turnham Green, by one highwayman, who despoiled the illustrious creature in sight of all his retinue; prisoners in London gaols fought battles with their turnkeys, and the majesty of the law fired blunderbusses in among them, loaded with rounds of shot and ball; thieves snipped off diamond crosses from the necks of noble lords at Court drawing-rooms; musketeers went into St. Giles's, to search for contraband goods, and the mob fired on the musketeers, and the musketeers fired on the mob, and nobody thought any of these occurrences much out of the common way. In the midst of them, the hangman, ever busy and ever worse than useless, was in constant requisition; now, stringing up long rows of miscellaneous criminals; now, hanging a housebreaker on Saturday who had been taken on Tuesday; now, burning people in the hand at Newgate by the dozen, and now burning pamphlets at the door of Westminster Hall; to-day, taking the life of an atrocious murderer, and to-morrow of a wretched pilferer who had robbed a farmer's boy of sixpence.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      "All these things, and a thousand like them, came to pass in and close upon the dear old year one thousand seven hundred and seventy-five. Environed by them, while the Woodman and the Farmer worked unheeded, those two of the large jaws, and those other two of the plain and the fair faces, trod with stir enough, and carried their divine rights with a high hand. Thus did the year one thousand seven hundred and seventy-five conduct their Greatnesses, and myriads of small creatures—the creatures of this chronicle among the rest—along the roads that lay before them.",
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      Object {
        "a": Object {
          "id": "link2H_4_0003",
        },
        "n": "a",
      },
    ],
    "n": "p",
  },
  Object {
    "c": Array [
      Object {
        "n": "br",
      },
      Object {
        "n": "br",
      },
      Object {
        "n": "br",
      },
      Object {
        "n": "br",
      },
    ],
    "n": "div",
  },
]
`;

exports[`convert({link}) 1`] = `
Array [
  Object {
    "a": Object {
      "epub:type": "toc",
      "id": "toc",
    },
    "c": Array [
      Object {
        "a": Object {
          "id": "toc-title",
        },
        "c": Array [
          "The Price of This Artist’s Work? A Conversation About the Horrors of War - The New York Times",
        ],
        "n": "h1",
      },
      Object {
        "c": Array [
          Object {
            "a": Object {
              "id": "toc-li-1",
            },
            "c": Array [
              Object {
                "a": Object {
                  "href": "ast/2.json#the-price-of-this-artists-work-a-conversation-about-the-horrors-of-war---the-new-york-times",
                },
                "c": Array [
                  "The Price of This Artist’s Work? A Conversation About the Horrors of War - The New York Times",
                ],
                "n": "a",
              },
            ],
            "n": "li",
          },
        ],
        "n": "ol",
      },
    ],
    "n": "nav",
  },
  Object {
    "a": Object {
      "epub:type": "landmarks",
      "hidden": "hidden",
      "id": "landmarks",
    },
    "c": Array [
      Object {
        "n": "ol",
      },
    ],
    "n": "nav",
  },
]
`;

exports[`nodeToAST() 1`] = `
Object {
  "c": Array [
    Object {
      "a": Object {
        "id": "h1",
      },
      "c": Array [
        "Lorem Ipsum",
      ],
      "n": "h1",
    },
    Object {
      "a": Object {
        "id": "h4",
      },
      "c": Array [
        " \\"Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...\\" ",
      ],
      "n": "h4",
    },
    Object {
      "c": Array [
        " \\"There is no one who loves pain itself, who seeks after it and wants to have it, simply because it is pain...\\" ",
      ],
      "n": "h5",
    },
    Object {
      "n": "hr",
    },
    Object {
      "c": Array [
        " Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed quis auctor erat. Sed accumsan, nibh non commodo dignissim, elit odio fringilla metus, ut ultrices turpis purus et orci. ",
        Object {
          "a": Object {
            "href": "https://example.com",
          },
          "c": Array [
            "Nullam posuere tempor dictum.",
          ],
          "n": "a",
        },
        " Donec nibh lectus, efficitur eu est quis, suscipit tempus lectus. Praesent eu sagittis libero, eu dapibus elit. Aliquam sed sapien magna. Fusce iaculis ultrices sapien, id fringilla nibh placerat eget. Phasellus cursus commodo velit sed condimentum. Praesent finibus euismod pharetra. Etiam vestibulum venenatis ligula. ",
      ],
      "n": "p",
    },
    Object {
      "c": Array [
        " Nunc lobortis posuere nulla, ",
        Object {
          "a": Object {
            "href": "lorem.html#h4",
          },
          "c": Array [
            "vel vestibulum",
          ],
          "n": "a",
        },
        " massa viverra vitae. Maecenas porttitor dui et malesuada maximus. Aenean non elementum risus, et luctus ex. Duis feugiat orci quam, a condimentum justo sagittis eget. Curabitur vulputate turpis vitae tellus convallis imperdiet. Vivamus imperdiet massa neque, aliquam feugiat odio molestie nec. Donec suscipit, felis sit amet scelerisque dapibus, leo nisi semper quam, id vehicula ligula purus in dui. Sed cursus diam ut metus egestas tempus. ",
      ],
      "n": "p",
    },
    Object {
      "c": Array [
        " Praesent lacinia risus elit, quis sollicitudin metus facilisis egestas. Proin id turpis est. Donec a malesuada felis. Proin molestie quis sem eget tincidunt. Vestibulum auctor consequat elit, eget mollis urna euismod ut. Aliquam suscipit justo nec turpis bibendum venenatis. Integer imperdiet dolor vitae erat hendrerit, ut tincidunt ligula convallis. Donec quis auctor ex, quis eleifend tellus. ",
      ],
      "n": "p",
    },
    Object {
      "c": Array [
        " Sed scelerisque mattis risus ut porta. Aenean condimentum ex nec lectus malesuada, et cursus ante vestibulum. Etiam consectetur felis lectus, quis consectetur sapien convallis eget. Morbi interdum mauris odio. Fusce et tempor mauris. Ut eget viverra ex. Sed tincidunt lacus aliquet justo consectetur posuere. Mauris vitae efficitur sapien, ut accumsan purus. Sed nulla diam, dignissim at cursus a, porttitor quis purus. Praesent sed neque ex. Suspendisse vitae suscipit nisi. Praesent vel eros hendrerit sapien posuere sollicitudin vitae sed sem. Integer commodo sit amet risus volutpat malesuada. Curabitur commodo velit eget aliquet eleifend. Donec ultrices mauris eu libero aliquam volutpat. ",
      ],
      "n": "p",
    },
    Object {
      "c": Array [
        " Fusce in leo nec lorem dapibus vulputate eu vitae tortor. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Integer aliquet, elit placerat semper egestas, dui risus fermentum felis, non molestie magna nibh eu mauris. Integer condimentum libero eu leo convallis porttitor. Proin at dapibus felis, eget volutpat ligula. Sed est tortor, ultrices sed leo in, ultricies tincidunt nibh. Nulla consectetur lobortis laoreet. Nunc faucibus, nulla sit amet aliquet faucibus, risus arcu aliquam libero, nec blandit massa felis sed odio. Vivamus tempus, ex nec feugiat bibendum, libero felis tristique lectus, eu vehicula magna quam ut augue. Nulla mattis purus turpis, vel suscipit metus pellentesque nec. ",
      ],
      "n": "p",
    },
  ],
  "n": "body",
}
`;
